src/Tasks/Microsoft.NET.Build.Tasks/GenerateShims.cs

```diff
@@ -0,0 +1,173 @@
+        /// </summary>
+        [Required]
+        public string TargetFrameworkMoniker { get; set; }
+
+        /// <summary>
+        /// PackageId of the dotnet tool NuGet Package.
+        /// </summary>
+        [Required]
+        public string PackageId { get; set; }
+
+        /// <summary>
+        /// Package Version of the dotnet tool NuGet Package.
+        /// </summary>
+        [Required]
+        public string PackageVersion { get; set; }
+
+        /// <summary>
+        /// The command name of the dotnet tool.
+        /// </summary>
+        [Required]
+        public string ToolCommandName { get; set; }
+
+        /// <summary>
+        /// The entry point of the dotnet tool which will be run by Apphost
+        /// </summary>
+        [Required]
+        public string ToolEntryPoint { get; set; }
+
+        /// <summary>
+        /// The output Directroy path of generated shims.
+        /// </summary>
+        [Required]
+        public string PackagedShimOutputDirectory { get; set; }
+
+        /// <summary>
+        /// The RuntimeIdentifiers that shims will be generated for.
+        /// </summary>
+        [Required]
+        public ITaskItem[] PackageToolShimRuntimeIdentifiers { get; set; }
+
+        /// <summary>
+        /// Path of generated shims. metadata "ShimRuntimeIdentifier" is used to map back to PackageToolShimRuntimeIdentifiers.
+        /// </summary>
+        [Output]
+        public ITaskItem[] EmbeddedApphostPaths { get; private set; }
+
+        protected override void ExecuteCore()
+        {
+            NuGetFramework targetFramework = NuGetUtils.ParseFrameworkName(TargetFrameworkMoniker);
+            LockFile lockFile = new LockFileCache(this).GetLockFile(ProjectAssetsFile);
+            _packageResolver = NuGetPackageResolver.CreateResolver(lockFile, ProjectPath);
+
+            var embeddedApphostPaths = new List<ITaskItem>();
+            foreach (var runtimeIdentifier in PackageToolShimRuntimeIdentifiers.Select(r => r.ItemSpec))
+            {
+                var resolvedApphostAssetPath = GetApphostAsset(targetFramework, lockFile, runtimeIdentifier);
+
+                var packagedShimOutputDirectoryAndRid = Path.Combine(PackagedShimOutputDirectory, runtimeIdentifier);
+                var appHostDestinationFilePath =
+                    Path.Combine(packagedShimOutputDirectoryAndRid,
+                        $"{ToolCommandName}{Path.GetExtension(resolvedApphostAssetPath)}");
+                var appBinaryFilePath =
+                    $".store/{PackageId.ToLowerInvariant()}/{PackageVersion}/{PackageId.ToLowerInvariant()}/{PackageVersion}/tools/{targetFramework.GetShortFolderName()}/any/{ToolEntryPoint}";
+
+                Directory.CreateDirectory(packagedShimOutputDirectoryAndRid);
+                if (File.Exists(appHostDestinationFilePath))
+                {
+                    File.Delete(appHostDestinationFilePath);
+                }
```

Comment:
This is because for other contexts, Apphost.Create assumes if host is already written, it doesn't need to be regenerated

I think it would be better to have a `bool overwriteExisting` parameter on AppHost.Create.

Just so I understand, is the issue that the args to Create can change without the path changing, hence we need to recreate it every time?

Reply:
Correct. I had a bug during my test. So I have to enforce recreate all the time. I will add `bool overwriteExisting` param to AppHost.Create